### ISSUE TITLE
Variable to specify number of zones in zone-aware deployments

### ISSUE THREAD

**Issue by kavfixnel** [grafana/mimir#5042]
#### Is your feature request related to a problem? Please describe. 

Add a variable to specify the number of zones in zone-aware deployments for stateful components in the Jsonnet library

#### Describe the solution you'd like

Instead of having a hardcoded number of zones for a/b/c, add logic to generate these zones with a loop and allow people to set the number of zones they want by changing a variable in the `$._config` object

**Comment by pracucci** [comment]
Hi @kavfixnel. I've the feeling this change will make jsonnet changes and overrides more complicated. Can you share more details about your use case? How many zone do you want to run? What's your replication factor?

**Comment by kavfixnel** [comment]
Yeah in my use case I want run 6 zones. The way I've successfully been able to do this is by setting the default [ingesters zone components to null](https://github.com/grafana/mimir/blob/main/operations/mimir/multi-zone.libsonnet#L128-L153) to null and creating a common object that has the ability to create an arbitrary number of zones through a loop:

```jsonnet
{
  _config: {
    num_ingester_zones: 3,
  },

  ingester_zone_a_container:: null,
  ingester_zone_a_statefulset: null,
  ingester_zone_a_service: null,

  ingester_zone_b_container:: null,
  ingester_zone_b_statefulset: null,
  ingester_zone_b_service: null,

  ingester_zone_c_container:: null,
  ingester_zone_c_statefulset: null,
  ingester_zone_c_service: null,
  
  ingester_zones: if !multi_zone_ingesters_deployed then null else {
    ['zone_%s' % std.char(97 + zone_num)]: {
      local zone_char = std.char(97 + zone_num),
      
      ingester_args:: {},
      ingester_env_map:: $.ingester_env_map,

      ingester_container::
        self.newIngesterZoneContainer(zone_char, $.ingester_zones['zone_%s' % zone_char].ingester_args, $.ingester_zones['zone_%s' % zone_char].ingester_env_map),

      ingester_statefulset:
        self.newIngesterZoneStatefulSet(zone_char, $.ingester_zones['zone_%s' % zone_char].ingester_container),

      ingester_service:
        $.newIngesterZoneService($.ingester_zones['zone_%s' % zone_char].ingester_statefulset),
    }
    for zone_num in std.range(0, $._config.num_ingester_zones - 1)
  }
}
```

This, of course, breaks compatibility, but I was not able to figure out how to coerce Jsonnet into dynamically creating these fields in the top level object without running into `Object comprehension can only have one field` error. I think if I knew how to "unpack" the object and create the `ingester_zone_x_[container|statefulset|service]` fields in the TLO to not break backwards compatibility it would be a great add and make make things more configurable.

This of course can also be applied to all zone-aware components.

**Comment by pstibrany** [comment]
Given the complexity of the change, I don't think we should support this use case.
